src/serac/physics/materials/thermal_expansion_material.cpp

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "serac/physics/materials/thermal_expansion_material.hpp"
+#include "serac/physics/utilities/physics_utils.hpp"
+
+#include "serac/infrastructure/logger.hpp"
+
+#include <cmath>
+
+namespace serac {
+
+inline void IsotropicThermalExpansionMaterial::EvalCoeffs() const
+{
+  coef_thermal_expansion_ = c_coef_thermal_expansion_->Eval(*parent_to_reference_transformation_,
+                                                            parent_to_reference_transformation_->GetIntPoint());
+  reference_temp_ =
+      c_reference_temp_->Eval(*parent_to_reference_transformation_, parent_to_reference_transformation_->GetIntPoint());
+  temp_ = temp_state_.gridFuncCoef().Eval(*parent_to_reference_transformation_,
+                                          parent_to_reference_transformation_->GetIntPoint());
+}
+
+void IsotropicThermalExpansionMaterial::modifyDisplacementGradient(mfem::DenseMatrix& du_dX)
+{
+  EvalCoeffs();
+
+  double expansion = coef_thermal_expansion_ * (reference_temp_ - temp_);
+
+  for (int i = 0; i < du_dX.Width(); ++i) {
+    du_dX(i, i) += expansion * (1.0 + du_dX(i, i));
+  }
+}
```

Comment:
Is this code doing the right thing? An isotropic scaling applied to the displacement gradient would affect all of its entries (not just diagonal ones). Or maybe this is actually calculating the thermal expansion deformation gradient, F_{\theta), in which case we should probably change the function/argument names.

Reply:
I believe this is correct. We commonly want material models to act on the displacement gradient, so that's why I'm modifying it here. My thinking is:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;\mathbf{F}_\textbf{total}&space;&=&space;\mathbf{F}_\textbf{thermal}\mathbf{F}_\textbf{mechanical}&space;\\&space;&=&space;(\mathbf{I}&plus;\mathbf{H}_\textbf{thermal})(\mathbf{I}&plus;\mathbf{H}_\textbf{mechanical})&space;\\&space;&=&space;\mathbf{I}&space;&plus;&space;\mathbf{H}_\textbf{thermal}&space;&plus;&space;\mathbf{H}_\textbf{mechanical}&space;&plus;&space;\mathbf{H}_\textbf{thermal}\mathbf{H}_\textbf{mechanical}&space;\\&space;&=&space;\mathbf{I}&space;&plus;&space;\mathbf{H}_\textbf{modified}&space;\\&space;\mathbf{H}_\textbf{thermal}&space;&=&space;\mathbf{F}_\textbf{thermal}&space;-&space;\mathbf{I}&space;\\&space;&=&space;\alpha&space;\left(&space;\theta&space;-&space;\theta_\textbf{ref}\right)\mathbf{I}&space;\end{aligh*}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;\mathbf{F}_\textbf{total}&space;&=&space;\mathbf{F}_\textbf{thermal}\mathbf{F}_\textbf{mechanical}&space;\\&space;&=&space;(\mathbf{I}&plus;\mathbf{H}_\textbf{thermal})(\mathbf{I}&plus;\mathbf{H}_\textbf{mechanical})&space;\\&space;&=&space;\mathbf{I}&space;&plus;&space;\mathbf{H}_\textbf{thermal}&space;&plus;&space;\mathbf{H}_\textbf{mechanical}&space;&plus;&space;\mathbf{H}_\textbf{thermal}\mathbf{H}_\textbf{mechanical}&space;\\&space;&=&space;\mathbf{I}&space;&plus;&space;\mathbf{H}_\textbf{modified}&space;\\&space;\mathbf{H}_\textbf{thermal}&space;&=&space;\mathbf{F}_\textbf{thermal}&space;-&space;\mathbf{I}&space;\\&space;&=&space;\alpha&space;\left(&space;\theta&space;-&space;\theta_\textbf{ref}\right)\mathbf{I}&space;\end{aligh*}" title="\begin{align*} \mathbf{F}_\textbf{total} &= \mathbf{F}_\textbf{thermal}\mathbf{F}_\textbf{mechanical} \\ &= (\mathbf{I}+\mathbf{H}_\textbf{thermal})(\mathbf{I}+\mathbf{H}_\textbf{mechanical}) \\ &= \mathbf{I} + \mathbf{H}_\textbf{thermal} + \mathbf{H}_\textbf{mechanical} + \mathbf{H}_\textbf{thermal}\mathbf{H}_\textbf{mechanical} \\ &= \mathbf{I} + \mathbf{H}_\textbf{modified} \\ \mathbf{H}_\textbf{thermal} &= \mathbf{F}_\textbf{thermal} - \mathbf{I} \\ &= \alpha \left( \theta - \theta_\textbf{ref}\right)\mathbf{I} \end{aligh*}" /></a>

Reply:
The last term in F_total is due to geometric nonlinearities. I was considering adding a flag to turn in off.

Reply:
sounds good, let's try to document this assumption (neglecting the product-term) in the doxygen comment for `modifyDisplacementGradient`, if possible